### PR TITLE
Add runner registration proxy support

### PR DESCRIPTION
**Pull request by alexjfisher**
A new `http_proxy` parameter can be set if you haven't got direct Internet access and need to use a proxy whilst registering runners with gitlab.  It is still the user's responsibility to configure other aspects of their runners to work with a proxy as needed.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -76,8 +76,13 @@ jobs:
...
       - name: Start gitlab
         run: ./scripts/start-gitlab.sh
+      - name: Start squid
+        run: ./scripts/start-squid.sh
       - name: Run tests
         run: bundle exec rake beaker
         env:
           BEAKER_PUPPET_COLLECTION: ${{ matrix.puppet.collection }}
           BEAKER_setfile: ${{ matrix.setfile.value }}
+      - name: squid logs
```
Comment (alexjfisher) [inline .github/workflows/ci.yml:86]: This was mainly for my debugging, but no harm in leaving it in.

File: REFERENCE.md [hunk 2]
```diff
@@ -210,6 +211,20 @@ The path to the config file of Gitlab runner.
...
+##### <a name="http_proxy"></a>`http_proxy`
+
+Data type: `Optional[Stdlib::HTTPUrl]`
+
+An HTTP proxy to use whilst registering runners.
+This setting is only used when registering runners and will be used for all runners in the `runners` parameter.
+If you have some runners that need to use a proxy and others that don't, leave `runners` and `http_proxy` unset and declare `gitlab_ci_runnner::runner` resources separately.
+If you do need to use an http proxy, you'll probably also want to configure other aspects of your runners to use it, (eg. setting `http_proxy` environment variables, `pre-clone-script`, `pre-build-script` etc.)
+Exactly how you might need to configure your runners varies between runner executors and specific use-cases.
+This module makes no attempt to automatically alter your runner configurations based on the value of this parameter.
```
Comment (alexjfisher) [inline REFERENCE.md:223]: I tried, then decided it was more harm than good.

File: lib/puppet/functions/gitlab_ci_runner/register_to_file.rb [hunk 3]
```diff
@@ -25,17 +26,19 @@
...
     param 'String[1]', :runner_name
     optional_param 'Hash', :additional_options
-    optional_param 'String[1]', :filename
+    optional_param 'Optional[String[1]]', :filename
+    optional_param 'Optional[String[1]]', :proxy
     return_type 'String[1]'
   end
 
-  def register_to_file(url, regtoken, runner_name, additional_options = {}, filename = "/etc/gitlab-runner/auth-token-#{runner_name}")
+  def register_to_file(url, regtoken, runner_name, additional_options = {}, filename = nil, proxy = nil)
```
Comment (alexjfisher) [inline lib/puppet/functions/gitlab_ci_runner/register_to_file.rb:34]: I'm wondering if `filename` really needs to be a parameter??

**Review by bastelfreak — APPROVED** [review]

File: lib/puppet/functions/gitlab_ci_runner/register_to_file.rb (hunk 3 above)
Comment (baurmatt) [reply lib/puppet/functions/gitlab_ci_runner/register_to_file.rb:34]: No, I think I only added it for debugging/convenience. Do you want to remove it?

File: manifests/init.pp [hunk 4]
```diff
@@ -43,7 +43,14 @@
 #   The keyserver which should be used to get the repository key.
 # @param config_path
 #   The path to the config file of Gitlab runner.
-#
+# @param http_proxy
+#   An HTTP proxy to use whilst registering runners.
+#   This setting is only used when registering runners and will be used for all runners in the `runners` parameter.
```
Comment (baurmatt) [inline manifests/init.pp:48]: Unregistering does use it as well :)

**Review by baurmatt — CHANGES_REQUESTED** [review]
Thanks, looks good! :) Just a small nit pick in the documentation :)

File: lib/puppet/functions/gitlab_ci_runner/register_to_file.rb (hunk 3 above)
Comment (alexjfisher) [reply lib/puppet/functions/gitlab_ci_runner/register_to_file.rb:34]: sure.  It'd make the code a little cleaner.

Comment (alexjfisher) [reply lib/puppet/functions/gitlab_ci_runner/register_to_file.rb]: updated

File: manifests/init.pp (hunk 4 above)
Comment (alexjfisher) [reply manifests/init.pp]: updated

**Pull request merged**